lib/IRGen/ClassMetadataVisitor.h

```diff
@@ -90,21 +90,31 @@ template <class Impl> class ClassMetadataVisitor
 private:
   /// Add fields associated with the given class and its bases.
   void addClassMembers(ClassDecl *theClass, Type type) {
-    // Add any fields associated with the superclass.
-    // NB: We don't apply superclass substitutions to members because we want
-    // consistent metadata layout between generic superclasses and concrete
-    // subclasses.
+    // Visit the superclass first.
     if (Type superclass = type->getSuperclass()) {
-      ClassDecl *superclassDecl = superclass->getClassOrBoundGenericClass();
-      // Skip superclass fields if superclass is resilient.
-      // FIXME: Needs runtime support to ensure the field offset vector is
-      // populated correctly.
-      if (!IGM.Context.LangOpts.EnableClassResilience ||
-          !IGM.isResilient(superclassDecl, ResilienceExpansion::Maximal)) {
+      auto *superclassDecl = superclass->getClassOrBoundGenericClass();
+      if (IGM.Context.LangOpts.EnableClassResilience &&
+          IGM.isResilient(superclassDecl, ResilienceExpansion::Maximal)) {
+        // Just note that we have a resilient superclass and move on.
+        //
+        // Runtime metadata instantiation needs to slide our entries down
+        // and copy in the superclass metadata.
+        //
+        // Metadata access needs to access our fields relative to a
+        // global variable.
+        asImpl().noteResilientSuperclass();
```

Comment:
I think we should note a resilient superclass even under minimal resilience, right?  If there's a module out there to which (1) this class might be exposed and (2) the superclass might be resilient, then this class needs to generate the structures to support resilient access (e.g. the start-of-metadata offset variable), even if this class isn't resilient.  We don't have to use those structures in code-generation, of course, so we may need to distinguish those cases.

It may make it easier to manage this kind of thing if the isResilient query actually returns the level of resilience instead of taking it as a parameter.

Reply:
The -enable-class-resilience flag is a staging flag and is going away.

Reply:
I'm not sure what you're getting at here; my comment is about the resilience-handling code you've added here, which will presumable still be active when the flag is folded to ``true``.

I'm trying to point out what I think is a bug: if the superclass is resilient to some module but not this one, and the subclass is not resilient, it looks like we won't create the offset variable.

Reply:
And I think that's an instance of a broader problem where you're doing layout based on Maximal to get optimal code-gen but also letting that affect what structures we generate, when structure-generation should generally be cued off of Minimal.

Reply:
I see what you mean now, I'll fix this and add tests in a follow-up commit. Thanks!

Reply:
Wait, the case you outlined cannot occur.

If the superclass is resilient to some other module but not this one, it means the superclass is defined inside this module and this module was built with resilience enabled.

This precludes the subclass being resilient, then, at least not as a supported configuration, because we don't want to productize `@fixedContents` on classes. While the standard library will need this for a couple of classes, none of them will have resilient base classes.

Note that this scenario does currently work:

- Superclass is resilient to us (defined in another resilient module)
- Subclass is not resilient because this module is not resilient

I'll add a test for this case. It does define the global and initialize it dynamically as expected.

Reply:
I mean, I would hope that we aren't leaving quiet potholes in the implementation based on random preconceptions of what sorts of resilience we're not going to productize.  Resilient superclass + non-resilient subclass does not seem like an impossible combination or one that is particularly time-saving to not support.